keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,639 @@
++ _COSI Controller_ - A central controller responsible for managing `Buckets`, `BucketContents`, and Secrets.
++ _Driver_ - A containerized gRPC server which implements a storage vendor’s business logic through the COSI interface. It can be written in any language supported by gRPC and is independent of Kubernetes.
++ _Greenfield Bucket_ - a new bucket created and managed by the COSI system.
++  _Object_ - An atomic, immutable unit of data stored in buckets.
++ _Sidecar_ - A `BucketContent` controller that communicates to the driver via a gRPC client.
++ _Static Bucket_ - externally created and manually integrated but _lacking_ a provisioner.
+
+# Proposal
+
+## User Stories
+
+#### Admin
+
+- As a cluster administrator, I can set quotas and resource limits on generated buckets' storage capacity via the Kubernete's API, so that  I can control monthly infrastructure costs.
+- As a cluster administrator, I can use Kubernetes RBAC policies on bucket APIs, so that I may control integration and access to pre-existing buckets from within the cluster, reducing the need to administer an external storage interface.
+- As a cluster administrator, I can manage multiple object store providers via the Kubernetes API, so that I do not have to become an expert in several different storage interfaces.
+
+#### User
+
+- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.
+- As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.
+
+  
+## System Configuration
+
++ The COSI controller runs in the `cosi-system` namespace where it manages `Buckets`, `BucketContents`, and Secrets. This namespace name is not enforced but suggested.
++ The Driver and Sidecar containers run together in a Pod and are deployed in any namespace, communicating via the Pod's internal network (_localhost:some-port_). We expect and will document that different drivers live in separate namespaces.
++ Operations must be idempotent in order to handle failure recovery.
+
+### Unique Driver Names
+
+It is important that driver names are unique, otherwise multiple sidecars would try to handle the same `BucketContent` events (since the sidecar matches on driver name).   The prescribed pattern  to be used for all provisioner names
+
+## Workflows
+
+
+
+### Determining Case from BucketClass
+
+| BucketClassFields             | SecretRef: non-nil | SecretRef: nil |
+| ----------------------------- | ------------------ | -------------- |
+| **BucketContentRef: non-nil** | Static             | Brownfield     |
+| **BucketContentRef: nil**     | Undefined          | Greenfield     |
+
+#### Create Bucket (Greenfield)
```

Comment:
`BucketContent` is confusing to me.  If you talk about the contents of a bucket, you're referring to the actual data objects held by the bucket. Similarly, you don't talk about a PV's content, you just talk about the PV. We should just talk about the "Bucket". What about these names?
- `BucketRequest` (or `BucketClaim`): The app's request for a bucket (renamed from `Bucket`)
- `Bucket`: Used for internally tracking the bucket creation/deletion (renamed from `BucketContent`)
- `BucketClass`: unchanged

Reply:
I (personally) like `BucketRequest` because that's how we often refer to it in conversations since "bucket" is ambiguous. I also like OBC but "claim" is not in favor with several folks. I prefer `BucketObject` or `OB` over `Bucket` due to the same ambiguous argument. When we say "bucket" it's sometimes hard to know if we mean the k8s resource or the physical bucket.